src/Database/LSMTree/Internal/RawPage.hs

```diff
@@ -47,6 +49,16 @@ data RawPage = RawPage
     !ByteArray
   deriving (Show)
 
+invariant :: RawPage -> Bool
+invariant (RawPage off ba) = and
+    [ off >= 0                                      -- offset is positive
+    , mod8 offsetInBytes == 0                       -- 64 bit/8 byte alignment
+    , (sizeofByteArray ba - offsetInBytes) >= 4096  -- there is always 4096 bytes
+    , isByteArrayPinned ba                          -- bytearray should be pinned
+    ]
+  where
+    offsetInBytes = mul2 off
```

Comment:
:+1:  Yes. Great.

Reply:
(I was referring to the invariant as a whole, but I didn't select the range of lines properly, oops.)